src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java

```diff
@@ -32,48 +38,96 @@ public RoleAssignmentsService(RoleService roleService) {
      * Remove from the user (at the caseload) the roles identified by specification.rolesToRemove.
      */
     @PreAuthorize("hasAnyRole('MAINTAIN_ACCESS_ROLES_ADMIN')")
-    public void updateRoleAssignments(@Valid final RoleAssignmentsSpecification specification) {
+    public Map<String, RoleAssignmentStats> updateRoleAssignments(@Valid final RoleAssignmentsSpecification specification) {
         log.info("Updating role assignments: {}", specification);
+        val caseloadResults = new TreeMap<String, RoleAssignmentStats>();
         for (val caseload : specification.getCaseloads()) {
 
             val usernamesForCaseload = findUsernamesMatchingRolesAtCaseload(specification.getRolesToMatch(), caseload);
             log.info("Found {} users for the {} caseload: {}.", usernamesForCaseload.size(), caseload, usernamesForCaseload);
-            usernamesForCaseload.forEach(username -> {
-                try {
-                    assignRolesToUser(username, specification.getRolesToAssign());
-                    try {
-                        removeRolesFromUserAtCaseload(caseload, username, specification.getRolesToRemove());
-                    } catch (Exception e) {
-                        val message = String.format("Failure while removing roles %1$s from user %2$s at caseload %3$s. Some roles may not have been removed. Continuing with next user.",specification.getRolesToRemove(), username, caseload);
-                        log.warn(message, e);
-                    }
-                } catch (Exception e) {
-                    val message = String.format("Failure while assigning roles %1$s to user %2$s. No roles will be removed from this user. Continuing with next user.", specification.getRolesToAssign(), username);
-                    log.warn(message, e);
+            val results = new RoleAssignmentStats();
+
+            for (val username : usernamesForCaseload) {
+                var success = assignRolesToUser(results, username, specification.getRolesToAssign());
```

Comment:
lots of mixing of `var` and `val` here.  My preference is `final var` since that's the only one that actually is final and also omits the type...

Reply:
stayed consistent with the existing code so used `val`s